### ISSUE TITLE
more examples

### ISSUE THREAD

**Issue by Gruski** [i8beef/Ecobee#3]
Is it possible for you to provide a few more API requests examples like you have with this:

```
var request = new ThermostatSummaryRequest  {
   Selection = new Protocol.Objects.Selection {
       SelectionType = "registered"
    }
};
var response = client.Get<ThermostatSummaryRequest, ThermostatSummaryResponse>(request).Result;
```

I need the following:
- get temperature of all sensors
- set a hold with a designated comfort zone
- set a sensor to participate in a comfort zone
- set system mode: heat, AC, auto, off
- set a hold with desired heat and cold temps

With just one single request as an example its really hard to map out the API docs to your code. You have no examples doing a POST with a JSON payload.

**Comment by i8beef** [comment]
Actually, there isn't much to document here... this is just a thin wrapper around the ecobee api, whIch they actually document pretty well. Those objects are direct pocos for serialization and the get and post methods are just raw http calls. The details you are looking for here on which data to passh and what to expect back are all provided by the ecobee api docs.

**Comment by Gruski** [comment]
Yes, I know and I looked extensively into ecobee's documentation. If it was obvious I wouldn't have asked. It is still not clear how to map the API to ecobee's documentation. I just need a few more examples if you don't mind especially one where you post a JSON payload. 

For example how to call this one:
https://www.ecobee.com/home/developer/api/examples/ex8.shtml

**Comment by i8beef** [comment]
Ok I'll walk you through their example:

1. https://www.ecobee.com/home/developer/api/examples/ex8.shtml - Says that you are going to POST a thermostat update request.
2. That is this object: https://github.com/i8beef/Ecobee/blob/master/src/I8Beef.Ecobee/Protocol/Thermostat/ThermostatUpdateRequest.cs

So you create a request like so:

```
var request = new ThermostatUpdateRequest {
    Selection = new Protocol.Objects.Selection { SelectionType = "registered" }
    Thermostat = new { Settings = new { HvacMode: "off" } }
}; 
var response = client.Post<ThermostatUpdateRequest, Response>(request).Result;
```

Obviously this is async, so the .Result is just unwrapping that in the test app, you should use this in a real await in production code.

You can see how I use some of these here as well: https://github.com/i8beef/HomeAutio.Mqtt.Ecobee/blob/master/src/HomeAutio.Mqtt.Ecobee/EcobeeMqttService.cs

The only real confusing part about some of these is that some of the requests make use of dynamic because you need to be able to send partial object structures to Ecobee for things like a ThermostatUpdateRequest, thus you need to read their docs and build the object correctly (no intellisense help there).

To help you map their docs to the object structures here, pretty much everything in "Protocol" maps to these request / response objects, namespaces map directly to the categories on the left there in their docs: https://www.ecobee.com/home/developer/api/documentation/v1/request-intro.shtml

The only odd ball is the "Objects" folder, and that directly maps again to these: https://www.ecobee.com/home/developer/api/documentation/v1/objects/api-object-definitions.shtml

All requests inherit from "RequestBase" and all responses inherit from "Response". "Reponse" is the simplest response Ecobee will send you, so it can be used as the response type for pretty much any call, but obviously you'd want to use the RIGHT Response class for say, a query requesting information, etc. In your example here, Ecobee's response to the request IS this simple response, (a lot of the fire and forget calls are), so thats why it's the type specified in the code example above.

**Comment by Gruski** [comment]
So I have this:

`        Dim objRequest = New ThermostatUpdateRequest With {
            .Functions = {
                New Protocol.Functions.SetHoldFunction With {
                    .Type = "setHold",
                    .Params = New Protocol.Functions.SetHoldParams With {
                        .HoldType = "indefinite",
                        .CoolHoldTemp = 900,
                        .HeatHoldTemp = 810
                    }
                }
            }
        }
        Dim objResponse = objClient.Post(Of ThermostatUpdateRequest, ThermostatResponse)(objRequest).Result`

**Comment by Gruski** [comment]
I am getting the Expect: 100-continue error. I saw your documentation on it. I did apply the servicePointManager XML entry in my app.config and error remains. My app.config looks like this:

```
<?xml version="1.0" encoding="utf-8" ?>
<configuration>
    <startup> 
        <supportedRuntime version="v4.0" sku=".NETFramework,Version=v4.6.1" />
    </startup>
    <system.net>
        <settings>
            <servicePointManager expect100Continue="false" />
        </settings>
    </system.net>
</configuration>
```

**Comment by Gruski** [comment]
Never mind on the 100-continue error. I put the XML in the inner app.config of my DLL instead of the calling windows service. Apparently it has to be in the outermost caller.

**Comment by Gruski** [comment]
I am not having any luck setting the temperature. Here is the code:

```
    Public Shared Function HoldTemp(strThermostatID As String, intHeat As Integer, intCool As Integer) As Task(Of ThermostatResponse)

        Dim objRequest = New ThermostatUpdateRequest With {
            .Selection = New Selection With {
                .SelectionType = "thermostats",
                .SelectionMatch = strThermostatID
            },
            .Functions = {
                New SetHoldFunction With {
                    .Params = New SetHoldParams With {
                        .HoldType = "indefinite",
                        .HeatHoldTemp = intHeat,
                        .CoolHoldTemp = intCool
                    }
                }
            }
        }
        Return objClient.Post(Of ThermostatUpdateRequest, ThermostatResponse)(objRequest)

    End Function
```
This is exactly how the documentation describes yet I get an error:
ApiException: 5: Invalid request format. Thermostat object in the request cannot be null.

I also see that in your code you wrote:
// TODO: Figure out how to pass both desired heat and cool values at same time

Ok since that is a bust I tried the other way of seting each temp separately as you did in your code:

```
    Public Shared Function HeatTemp(strThermostatID As String, intTemp As Integer) As Task(Of ThermostatResponse)

        Dim objRequest = New ThermostatUpdateRequest With {
            .Selection = New Selection With {
                .SelectionType = "thermostats",
                .SelectionMatch = strThermostatID
            },
            .Thermostat = New With {
                .Runtime = New With {
                    .DesiredHeat = intTemp
                }
            }
        }
        Return objClient.Post(Of ThermostatUpdateRequest, ThermostatResponse)(objRequest)

    End Function
```
Now I get: ApiException: 4: Serialization error. Update of property Thermostat.runtime is denied.
Sure enough the documentation states that the DesiredHeat field is read only.

So how in the world do you do a simple thing as setting the temperature?

**Comment by Gruski** [comment]
How does one use the Climate protocol? Climate is in the Program object but the ThermostatUpdateRequest has no Program member. So how does one go about assigning remote sensors to a climate?

**Comment by i8beef** [comment]
Your questions here are largely around how their API works, not details of this library. Their documentation is actually pretty decent here, and you are going to have to explore a bit. I only use certain pieces here right now personally, and it looks like I never finished the desired heat / cool stuff in my app using this library actually. I suspect you are going to have to set a HOLD (temporary) or modify the PROGRAM (permanent) to write new temperature values, but you'll have to read their docs to figure out how they want you to do it.

The "Thermostats null" issue though might be a legitimate bug here in my side. Because of their use of partial object graphs, I suspect the JSON serializer needs to be set to ignore NULL objects during serialization. I'll change that tomorrow and push a new version. I'll also upgrade the library for a few other pieces I need to do here (Async suffixes where appropriate, better HttpClient lifetime management, .NET Standard 2.0 support).

**Comment by Perecli** [comment]
Do you know why I might be getting this with your new version when I run your library code from my test project?
Could not load file or assembly 'Newtonsoft.Json, Version=11.0.0.0,

**Comment by i8beef** [comment]
Ok, you can try the updated branch 

https://github.com/i8beef/Ecobee/tree/1-NetStandardOverhaul

I'll be releasing this as a 2.0 here this weekend after I do a little smoke testing. The Async implementation is standardized so the calls will change to be suffixed with "Async", but you should be able to do the following now:

1. Drop the service point config piece. It should handle this internally now.
2. Your thermostat function call without a thermostats object should now work as the serializer is changed to ignore null objects.

**Comment by i8beef** [comment]
@Bortxx Are you just running the sample app or are you trying to reference the compiled library from some other application? If its the latter, because I updated this all to .NET Standard 2.0, your app will need to be 4.6.1 or later, or Core 2.0. Also, you'll need to make sure your app is referencing the latest JSON.NET here as the library reference was updated.

**Comment by Perecli** [comment]
So I have may own library that references yours and then I have a test harness that references my library. Both my projects use .NET Framework 4.6.1

What is JSON.NET? Do you mean I have to use nuget to get Newtonsoft and reference it from both my projects or just my library project?
Or are you talking about System.Runtime.Serialization.Json.dll?

And how come your TestClient project doesn't need a reference to Newtonsoft and works fine?

**Comment by i8beef** [comment]
Because its direct referencing from inside the same solution and you are referencing the DLL as a file. VS knows how to walk the dependency tree in the former to pick up the dependencies of the other project, but is reliant on you providing the dependencies correctly in the latter.

JSON.NET is basically the standard for JSON serialization in the .NET ecosystem. Yes, you can just install it via NuGet. When this is published as a NuGet package this weekend, NuGet will take care of all that for you without you manually needing to install it.

**Comment by Perecli** [comment]
Mine was direct referencing inside the same solution too. I had them all in one.

What I ended up doing is made my lib project a .NET standard 2.0 and made my test harness a .NET core app like yours instead of using .NET Framework 4.6.1 and that fixed the Newtonsoft conflict issue.

After I did that I tested the temp setting call and it worked (error went away).

Thank you for doing this and helping me out. BTW, I am the same user as Gruski. That is my work account.
